src/fe_utils/frontend_hook.c

```diff
@@ -0,0 +1,166 @@
+#include "fe_utils/frontend_hook.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#include <dlfcn.h>    // for dlopen()
+#include <sys/stat.h> // for stat()
+
+#include "c.h"
+#include "port.h"
+#include "pg_config.h"
+
+#define MAXPATHLEN 256 // in src/port/glob.c
+
+char FrontednHookPgInstallPath[MAXPATHLEN] = "\0", FrontendHookPgDataPath[MAXPATHLEN] = "\0";
+
+typedef struct df_files {
+	struct df_files *next;    /* List link */
+	void            *handle;  /* a handle for pg_dl* functions */
+	struct stat     fileinfo; /* device id and inode */
+	char            full_path[FLEXIBLE_ARRAY_MEMBER];
+} DynamicFileList;
+
+/* same as server, do not call dlclose for loaded libraries, check internal_unload_library before version 14 */
+static DynamicFileList *file_list = NULL;
+
+// find which libraries need to load
+static void fronted_load_library(const char *full_path) {
+	DynamicFileList *file_scanner;
+
+	/*
+	 * Scan the list of loaded FILES to see if the file has been loaded.
+	 */
+	for (file_scanner = file_list;
+		file_scanner != NULL &&
+		strcmp(full_path, file_scanner->full_path) != 0;
+		file_scanner = file_scanner->next)
+	{
+	}
+
+	if (file_scanner == NULL)
+	{
+		/*
+		 * File does not loaded yet.
+		 */
+		struct stat stat_buf;
+		if (lstat(full_path, &stat_buf) != 0)
+		{
+			printf("lstat: %s\n", strerror(errno));
+			exit(-1);
+		}
+
+		void *h = dlopen(full_path, RTLD_NOW | RTLD_GLOBAL);
+		if (h == NULL)
+		{
+			char *error = dlerror();
+			printf("dlopen: %s\n", error);
+			exit(-1);
+		}
+
+		void (*f)(void) = dlsym(h, "_PG_init");
+		if (f == NULL)
+		{
+			printf("dlopen: can not find _PG_init() in %s\n", full_path);
+			dlclose(h);
+			exit(-1);
+		}
+
+		(void)f();
+
+		DynamicFileList *current = malloc(sizeof(DynamicFileList) + strlen(full_path));
```

Comment:
`strlen(full_path) + 1`

https://github.com/postgres/postgres/blob/9e72f6bfae1707fca1a297db433f6145d582a1c2/src/backend/utils/fmgr/dfmgr.c#L225

I know you have carefully used `strncpy` to ensure there is no over-written of the target memory. However, when do the `strcmp` at line 40, the memory could still be "over-read", which is a bit "undefined" behavior.